### PR TITLE
 Don't trigger while_let_on_iterator when the iterator is recreated every iteration

### LINKED ISSUE #1654: while_let_on_iterator false positive
Also from [`toml-rs`](https://github.com/alexcrichton/toml-rs/pull/157/files#diff-d6f0a2686a3a16874359dca78545065bL232):

```
while let Some((_, ch)) = self.chars.clone().next() {
```

This should not be changed to:

```
for (_, ch) in self.chars.clone() {
```

Because they are semantically different.

### PR DESCRIPTION
**Pull request by flip1995**
r? @phansch 

Fixes #1654

changelog: Fix false positive in [`while_let_on_iterator`]

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -566,6 +566,17 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Loops {
             ) = (pat, &match_expr.kind)
             {
                 let iter_expr = &method_args[0];
+
+                // Don't lint when the iterator is recreated on every iteration
+                if_chain! {
+                    if let ExprKind::MethodCall(..) | ExprKind::Call(..) = iter_expr.kind;
+                    if let Some(iter_def_id) = get_trait_def_id(cx, &paths::ITERATOR);
+                    if implements_trait(cx, cx.tables.expr_ty(iter_expr), iter_def_id, &[]);
```
Comment (phansch) [inline clippy_lints/src/loops.rs:574]: Just to make sure I understand correctly: this is checking that the scrutinee is not calling an additional `iter` method, right?

Comment (flip1995) [reply clippy_lints/src/loops.rs:574]: It is checking, that a (method) call before the `next` call does not return a type, that implements iterator. Should I add a comment?

This does not cover this:

```rust
#[derive(Debug, Clone)]
struct S<'a> {
    chars: std::str::CharIndices<'a>,
}

while let Some(_) = s.clone().chars.next() {
    break;
}
```

But then again, if the struct is clonable, every field is clonable (in general), and this should be changed to `s.chars.clone().next()`, which prevents the lint from triggering.

**Pull request merged**